### PR TITLE
[Flaky] Integration tests - Using cohorts for sharing unused resources 

### LINKED ISSUE #3633: [Flaky test] Using cohorts for sharing unused resources Should start workloads that are under ...
/kind flake

**What happened**:

Failure https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_kueue/3632/pull-kueue-test-integration-main/1861018475376414720

**What you expected to happen**:

No failures

**How to reproduce it (as minimally and precisely as possible)**:

Repeat on CI

**Anything else we need to know?**:

```
{Timed out after 5.001s.
The function passed to Eventually failed at /home/prow/go/src/sigs.k8s.io/kueue/test/util/util.go:248 with:
Not enough workloads were admitted
Expected
    <int>: 0
to equal
    <int>: 1 failed [FAILED] Timed out after 5.001s.
The function passed to Eventually failed at /home/prow/go/src/sigs.k8s.io/kueue/test/util/util.go:248 with:
Not enough workloads were admitted
Expected
    <int>: 0
to equal
    <int>: 1
In [It] at: /home/prow/go/src/sigs.k8s.io/kueue/test/integration/scheduler/scheduler_test.go:1271 @ 11/25/24 12:18:31.598
}
```

### PR DESCRIPTION
**Pull request by mszadkow**
#### What type of PR is this?
/kind flake

#### What this PR does / why we need it:
Make sure previous workload has finished so cq has enough resources for another workload.
To avoid flaky test.
#### Which issue(s) this PR fixes:

Fixes #3633

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: test/integration/scheduler/scheduler_test.go [hunk 1]
```diff
@@ -1267,12 +1267,14 @@ var _ = ginkgo.Describe("Scheduler", func() {
 
 			ginkgo.By("Finishing second workload for prod ClusterQueue")
 			util.FinishWorkloads(ctx, k8sClient, pWl2)
+			util.ExpectWorkloadToFinish(ctx, k8sClient, client.ObjectKeyFromObject(pWl2))
```
Comment (mbobrovskyi) [inline test/integration/scheduler/scheduler_test.go:1270]: Not sure I got how it can help.

Comment (mszadkow) [reply test/integration/scheduler/scheduler_test.go:1270]: Yes, we set the status, but we do not wait for the status to be actually changed before we continue

Comment (mszadkow) [reply test/integration/scheduler/scheduler_test.go:1270]: I have tested it in 3 ways:
1. repeat 10 times with no changes - couple failed bc of the issue
2. repeat 10 times with extended time, like 5* the current timeout - works all the time
3. repeat 10 times with this fix and timeout stays as it was in 1) - works all the time

Comment (mimowo) [reply test/integration/scheduler/scheduler_test.go:1270]: I see, so it basically splits the 5s timeout into two phases: wait for workload to finish, and wait for wl3 to reserve quota. I guess the test fails if both things don't happen within 5s currently.

**Comment by mszadkow** [comment]
Comparison of successful and failed test run:

Success
```bash
STEP: Finishing second workload for prod ClusterQueue @ 12/05/24 12:33:30.03
  2024-12-05T12:33:30.039455+01:00      LEVEL(-2)       workload-reconciler     core/workload_controller.go:681 Workload update event   {"workload": {"name":"p-wl-2","namespace":"core-cxq98"}, "queue": "prod-queue", "status": "finished", "prevStatus": "admitted", "prevClusterQueue": "prod-cq"}
  2024-12-05T12:33:30.039603+01:00      LEVEL(-2)       workload-reconciler     queue/manager.go:489    Attempting to move workloads    {"workload": {"name":"p-wl-2","namespace":"core-cxq98"}, "queue": "prod-queue", "status": "finished", "cohort": "all", "root": "all"}
  2024-12-05T12:33:30.039752+01:00      LEVEL(-2)       core/workload_controller.go:143 Reconciling Workload    {"controller": "workload", "controllerGroup": "kueue.x-k8s.io", "controllerKind": "Workload", "Workload": {"name":"p-wl-2","namespace":"core-cxq98"}, "namespace": "core-cxq98", "name": "p-wl-2", "reconcileID": "553dd96a-3023-481d-9a3d-da60c21fd094", "workload": {"name":"p-wl-2","namespace":"core-cxq98"}}
  2024-12-05T12:33:30.110434+01:00      LEVEL(-3)       scheduler       queue/manager.go:531    Obtained ClusterQueue heads     {"attemptCount": 13, "count": 2}
  2024-12-05T12:33:30.110767+01:00      LEVEL(-2)       scheduler       scheduler/scheduler.go:543      Workload assumed in the cache   {"attemptCount": 13, "workload": {"name":"p-wl-3","namespace":"core-cxq98"}, "clusterQueue": {"name":"prod-cq"}}
  2024-12-05T12:33:30.110836+01:00      LEVEL(-3)       scheduler       scheduler/logging.go:43 Workload evaluated for admission        {"attemptCount": 13, "workload": {"name":"p-wl-3","namespace":"core-cxq98"}, "clusterQueue": {"name":"prod-cq"}, "status": "assumed", "reason": ""}
  2024-12-05T12:33:30.110891+01:00      LEVEL(-3)       scheduler       scheduler/logging.go:43 Workload evaluated for admission        {"attemptCount": 13, "workload": {"name":"d-wl-1","namespace":"core-cxq98"}, "clusterQueue": {"name":"dev-cq"}, "status": "skipped", "reason": "Workload no longer fits after processing another workload"}
  2024-12-05T12:33:30.110982+01:00      LEVEL(-2)       scheduler       scheduler/scheduler.go:636      Workload re-queued      {"attemptCount": 13, "workload": {"name":"d-wl-1","namespace":"core-cxq98"}, "clusterQueue": {"name":"dev-cq"}, "queue": {"name":"dev-queue","namespace":"core-cxq98"}, "requeueReason": "FailedAfterNomination", "added": true, "status": "skipped"}
```

Fail
```bash
  STEP: Finishing second workload for prod ClusterQueue @ 12/05/24 12:33:46.336
  2024-12-05T12:33:46.342581+01:00      LEVEL(-3)       scheduler       queue/manager.go:531    Obtained ClusterQueue heads     {"attemptCount": 12, "count": 2}
  2024-12-05T12:33:46.342903+01:00      LEVEL(-2)       scheduler       scheduler/scheduler.go:543      Workload assumed in the cache   {"attemptCount": 12, "workload": {"name":"d-wl-1","namespace":"core-c628d"}, "clusterQueue": {"name":"dev-cq"}}
  2024-12-05T12:33:46.342956+01:00      LEVEL(-2)       scheduler       scheduler/scheduler.go:228      Workload requires preemption, but there are no candidate workloads allowed for preemption   {"attemptCount": 12, "workload": {"name":"p-wl-3","namespace":"core-c628d"}, "clusterQueue": {"name":"prod-cq"}, "preemption": {"reclaimWithinCohort":"Never","borrowWithinCohort":{"policy":"Never"},"withinClusterQueue":"Never"}}
  2024-12-05T12:33:46.34301+01:00       LEVEL(-3)       scheduler       scheduler/logging.go:43 Workload evaluated for admission        {"attemptCount": 12, "workload": {"name":"d-wl-1","namespace":"core-c628d"}, "clusterQueue": {"name":"dev-cq"}, "status": "assumed", "reason": ""}
  2024-12-05T12:33:46.343057+01:00      LEVEL(-3)       scheduler       scheduler/logging.go:43 Workload evaluated for admission        {"attemptCount": 12, "workload": {"name":"p-wl-3","namespace":"core-c628d"}, "clusterQueue": {"name":"prod-cq"}, "status": "", "reason": "couldn't assign flavors to pod set main: insufficient unused quota for cpu in flavor on-demand, 1 more needed"}
  2024-12-05T12:33:46.343146+01:00      LEVEL(-2)       scheduler       scheduler/scheduler.go:636      Workload re-queued      {"attemptCount": 12, "workload": {"name":"p-wl-3","namespace":"core-c628d"}, "clusterQueue": {"name":"prod-cq"}, "queue": {"name":"prod-queue","namespace":"core-c628d"}, "requeueReason": "", "added": true, "status": ""}
  2024-12-05T12:33:46.344572+01:00      LEVEL(-2)       workload-reconciler     core/workload_controller.go:681 Workload update event   {"workload": {"name":"p-wl-2","namespace":"core-c628d"}, "queue": "prod-queue", "status": "finished", "prevStatus": "admitted", "prevClusterQueue": "prod-cq"}
  2024-12-05T12:33:46.344729+01:00      LEVEL(-2)       workload-reconciler     queue/manager.go:489    Attempting to move workloads    {"workload": {"name":"p-wl-2","namespace":"core-c628d"}, "queue": "prod-queue", "status": "finished", "cohort": "all", "root": "all"}
  2024-12-05T12:33:46.344934+01:00      LEVEL(-2)       core/workload_controller.go:143 Reconciling Workload    {"controller": "workload", "controllerGroup": "kueue.x-k8s.io", "controllerKind": "Workload", "Workload": {"name":"p-wl-2","namespace":"core-c628d"}, "namespace": "core-c628d", "name": "p-wl-2", "reconcileID": "ae2e1b24-84a8-481c-88f4-de04deb9db09", "workload": {"name":"p-wl-2","namespace":"core-c628d"}}
```

What we can see here (this is all without the change) is that:
Successful run starts with `p-wl-2` going from `admitted` to `finish`.
Then workloads are evaluated for admission and `p-wl-3` is assumed.

In failed run first workloads are evaluated fro admission and `d-wl-1` gets assumed...
So the order is reversed, and `d-wl-1` is assumed bc there is enough resources for it (1 CPU released already from `p-wl-1`).

**Comment by mszadkow** [comment]
> Comparison of successful and failed test run:
> 
> Success
> 
> ```shell
> STEP: Finishing second workload for prod ClusterQueue @ 12/05/24 12:33:30.03
>   2024-12-05T12:33:30.039455+01:00      LEVEL(-2)       workload-reconciler     core/workload_controller.go:681 Workload update event   {"workload": {"name":"p-wl-2","namespace":"core-cxq98"}, "queue": "prod-queue", "status": "finished", "prevStatus": "admitted", "prevClusterQueue": "prod-cq"}
>   2024-12-05T12:33:30.039603+01:00      LEVEL(-2)       workload-reconciler     queue/manager.go:489    Attempting to move workloads    {"workload": {"name":"p-wl-2","namespace":"core-cxq98"}, "queue": "prod-queue", "status": "finished", "cohort": "all", "root": "all"}
>   2024-12-05T12:33:30.039752+01:00      LEVEL(-2)       core/workload_controller.go:143 Reconciling Workload    {"controller": "workload", "controllerGroup": "kueue.x-k8s.io", "controllerKind": "Workload", "Workload": {"name":"p-wl-2","namespace":"core-cxq98"}, "namespace": "core-cxq98", "name": "p-wl-2", "reconcileID": "553dd96a-3023-481d-9a3d-da60c21fd094", "workload": {"name":"p-wl-2","namespace":"core-cxq98"}}
>   2024-12-05T12:33:30.110434+01:00      LEVEL(-3)       scheduler       queue/manager.go:531    Obtained ClusterQueue heads     {"attemptCount": 13, "count": 2}
>   2024-12-05T12:33:30.110767+01:00      LEVEL(-2)       scheduler       scheduler/scheduler.go:543      Workload assumed in the cache   {"attemptCount": 13, "workload": {"name":"p-wl-3","namespace":"core-cxq98"}, "clusterQueue": {"name":"prod-cq"}}
>   2024-12-05T12:33:30.110836+01:00      LEVEL(-3)       scheduler       scheduler/logging.go:43 Workload evaluated for admission        {"attemptCount": 13, "workload": {"name":"p-wl-3","namespace":"core-cxq98"}, "clusterQueue": {"name":"prod-cq"}, "status": "assumed", "reason": ""}
>   2024-12-05T12:33:30.110891+01:00      LEVEL(-3)       scheduler       scheduler/logging.go:43 Workload evaluated for admission        {"attemptCount": 13, "workload": {"name":"d-wl-1","namespace":"core-cxq98"}, "clusterQueue": {"name":"dev-cq"}, "status": "skipped", "reason": "Workload no longer fits after processing another workload"}
>   2024-12-05T12:33:30.110982+01:00      LEVEL(-2)       scheduler       scheduler/scheduler.go:636      Workload re-queued      {"attemptCount": 13, "workload": {"name":"d-wl-1","namespace":"core-cxq98"}, "clusterQueue": {"name":"dev-cq"}, "queue": {"name":"dev-queue","namespace":"core-cxq98"}, "requeueReason": "FailedAfterNomination", "added": true, "status": "skipped"}
> ```
> 
> Fail
> 
> ```shell
>   STEP: Finishing second workload for prod ClusterQueue @ 12/05/24 12:33:46.336
>   2024-12-05T12:33:46.342581+01:00      LEVEL(-3)       scheduler       queue/manager.go:531    Obtained ClusterQueue heads     {"attemptCount": 12, "count": 2}
>   2024-12-05T12:33:46.342903+01:00      LEVEL(-2)       scheduler       scheduler/scheduler.go:543      Workload assumed in the cache   {"attemptCount": 12, "workload": {"name":"d-wl-1","namespace":"core-c628d"}, "clusterQueue": {"name":"dev-cq"}}
>   2024-12-05T12:33:46.342956+01:00      LEVEL(-2)       scheduler       scheduler/scheduler.go:228      Workload requires preemption, but there are no candidate workloads allowed for preemption   {"attemptCount": 12, "workload": {"name":"p-wl-3","namespace":"core-c628d"}, "clusterQueue": {"name":"prod-cq"}, "preemption": {"reclaimWithinCohort":"Never","borrowWithinCohort":{"policy":"Never"},"withinClusterQueue":"Never"}}
>   2024-12-05T12:33:46.34301+01:00       LEVEL(-3)       scheduler       scheduler/logging.go:43 Workload evaluated for admission        {"attemptCount": 12, "workload": {"name":"d-wl-1","namespace":"core-c628d"}, "clusterQueue": {"name":"dev-cq"}, "status": "assumed", "reason": ""}
>   2024-12-05T12:33:46.343057+01:00      LEVEL(-3)       scheduler       scheduler/logging.go:43 Workload evaluated for admission        {"attemptCount": 12, "workload": {"name":"p-wl-3","namespace":"core-c628d"}, "clusterQueue": {"name":"prod-cq"}, "status": "", "reason": "couldn't assign flavors to pod set main: insufficient unused quota for cpu in flavor on-demand, 1 more needed"}
>   2024-12-05T12:33:46.343146+01:00      LEVEL(-2)       scheduler       scheduler/scheduler.go:636      Workload re-queued      {"attemptCount": 12, "workload": {"name":"p-wl-3","namespace":"core-c628d"}, "clusterQueue": {"name":"prod-cq"}, "queue": {"name":"prod-queue","namespace":"core-c628d"}, "requeueReason": "", "added": true, "status": ""}
>   2024-12-05T12:33:46.344572+01:00      LEVEL(-2)       workload-reconciler     core/workload_controller.go:681 Workload update event   {"workload": {"name":"p-wl-2","namespace":"core-c628d"}, "queue": "prod-queue", "status": "finished", "prevStatus": "admitted", "prevClusterQueue": "prod-cq"}
>   2024-12-05T12:33:46.344729+01:00      LEVEL(-2)       workload-reconciler     queue/manager.go:489    Attempting to move workloads    {"workload": {"name":"p-wl-2","namespace":"core-c628d"}, "queue": "prod-queue", "status": "finished", "cohort": "all", "root": "all"}
>   2024-12-05T12:33:46.344934+01:00      LEVEL(-2)       core/workload_controller.go:143 Reconciling Workload    {"controller": "workload", "controllerGroup": "kueue.x-k8s.io", "controllerKind": "Workload", "Workload": {"name":"p-wl-2","namespace":"core-c628d"}, "namespace": "core-c628d", "name": "p-wl-2", "reconcileID": "ae2e1b24-84a8-481c-88f4-de04deb9db09", "workload": {"name":"p-wl-2","namespace":"core-c628d"}}
> ```
> 
> What we can see here (this is all without the change) is that: Successful run starts with `p-wl-2` going from `admitted` to `finish`. Then workloads are evaluated for admission and `p-wl-3` is assumed.
> 
> In failed run first workloads are evaluated fro admission and `d-wl-1` gets assumed... So the order is reversed, and `d-wl-1` is assumed bc there is enough resources for it (1 CPU released already from `p-wl-1`).

I would like to discuss the name of the test:
Should start workloads that are under min quota before borrowing

I don't see the special setup with borrowing or lending quotas here, maybe that could help here, but now it's just how it works.
wdyt @mbobrovskyi ?

**Comment by mimowo** [comment]
cc @gabesaba

**Comment by mbobrovskyi** [comment]
> > What we can see here (this is all without the change) is that: Successful run starts with `p-wl-2` going from `admitted` to `finish`. Then workloads are evaluated for admission and `p-wl-3` is assumed.
> > In failed run first workloads are evaluated fro admission and `d-wl-1` gets assumed... So the order is reversed, and `d-wl-1` is assumed bc there is enough resources for it (1 CPU released already from `p-wl-1`).
> 
> I would like to discuss the name of the test: Should start workloads that are under min quota before borrowing
> 
> I don't see the special setup with borrowing or lending quotas here, maybe that could help here, but now it's just how it works. wdyt @mbobrovskyi ?

Yeah, that what I thought. dWL1 admitted before pWL3. for some reason.

File: test/integration/scheduler/scheduler_test.go [hunk 2]
```diff
@@ -1267,12 +1267,14 @@ var _ = ginkgo.Describe("Scheduler", func() {
...
 			// The pWl3 workload gets accepted, even though it was created after dWl1.
```
Comment (mbobrovskyi) [inline test/integration/scheduler/scheduler_test.go:1271]: Also please use `util.WaitForNextSecondAfterCreation(dWl1)` after dWL1 created and before pWl3 created to implement this case..

File: test/integration/scheduler/scheduler_test.go (hunk 1 above)
Comment (mimowo) [reply test/integration/scheduler/scheduler_test.go:1270]: @mszadkow is my understanding above correct - just giving more time, or something more is going on? If this is not just about more time, then I also still don't understand why the fix helps.

Comment (mbobrovskyi) [reply test/integration/scheduler/scheduler_test.go:1270]: I don't think these changes help. I think it's just a coincidence.

**Review by mimowo — COMMENTED** [review]
/approve

**Comment by mimowo** [comment]
/cherry-pick release-0.9
To make the release branch more stable

**Pull request merged**